library("tidyverse")
library(chromVAR)
library(GenomicRanges)
library(here)
library(chromVARmotifs)
library(motifmatchr)
library(Matrix)
library(SummarizedExperiment)
library(BiocParallel)
set.seed(2017)
register(MulticoreParam(4, progressbar = TRUE))
#BiocManager::install("BSgenome.Hsapiens.UCSC.hg38", version = "3.8")
library(BSgenome.Hsapiens.UCSC.hg38)

sampleMetadata   <- read_tsv(here('extractedData' , 'sampleMetadata_SI2-SI4.txt'), col_names=TRUE)
relevantMetadata <- filter(sampleMetadata, between(as.integer(substr(sampleID,1,2)),1,36))

#load the fragmentCounts object
fragmentCountsDiffPeaks <- read_rds(here('extractedData', 'atacFragmentCountsAllCondsDifferentialPeaks.rds'))
fragmentCountsDiffPeaks <- addGCBias(fragmentCountsDiffPeaks, 
                                     genome = BSgenome.Hsapiens.UCSC.hg38)
data("human_pwms_v2") #loads the curated cisBP motif set from the chromVar paper
motifSet <- human_pwms_v2 #loads the curated cisBP motif set from the chromVar paper
motif_ix <- matchMotifs(motifSet, fragmentCountsDiffPeaks, 
                        genome = BSgenome.Hsapiens.UCSC.hg38)
dev <- computeDeviations(object = fragmentCountsDiffPeaks, annotations = motif_ix)
variability <- computeVariability(dev)
plotVariability(variability, use_plotly = FALSE)
tvar <- as.tibble(variability)
# geneNamesTibble <- read_tsv('/Users/emsanford/Dropbox (RajLab)/Shared_Eric/HistoryDependence/SequencingResults/Analysis_HD3_to_HD6/refs/EnsgToHg38CanonicalTssAndHgncSymbol.tsv', col_names = TRUE)
# #dds is generated by deseq2
# geneNamesTibble <- tibble(gene_id = rownames(rowData(dds)), baseMean = rowData(dds)$baseMean)
# deseqBaseMeanTibbleWithHGNC <- left_join(deseqBaseMeanTibble, geneNamesTibble, by = 'gene_id')


#motifSet <- getJasparMotifs()  #loads the 2016 JASPAR CORE motif set

minVarThresh <- 6
minVariabilityZscoreNames <- rownames(variability[variability$variability > minVarThresh,])
devSubset <- dev[minVariabilityZscoreNames,]
dev.zscores <- assays(devSubset)[[2]]
forpca <- t(dev.zscores)
pcaresults <- prcomp(forpca, center = TRUE)
pcloadings <- as.matrix(pcaresults[[2]])
pcscores <- as.matrix(forpca) %*% pcloadings
pcanames <- rownames(forpca)
pcscorestibble <- as.tibble(pcscores)
pctVarExplained <- pcaresults$sdev^2 / sum(pcaresults$sdev^2)


sampleNames <- as_vector(lapply(strsplit(colnames(fragmentCountsDiffPeaks), '-rep'), function(x) x[1]))
conditions <- left_join(tibble(sampleID = sampleNames), relevantMetadata)$condition

# plot conditions on their PCA axes
ggplot(data=pcscorestibble) + 
  geom_point(mapping=aes(x=pcscores[,1], y= pcscores[,2], color=conditions)) + 
  geom_text(mapping=aes(x=pcscores[,1], y=pcscores[,2], label=sampleNames), size=3, nudge_y = 8, check_overlap = FALSE, angle=0) + 
  ggtitle('PCA of motif deviation z-scores in chromVar') + 
  xlab(sprintf('PC1: %.2f pct variance ', pctVarExplained[1]*100)) + 
  ylab(sprintf('PC2: %.2f\ pct variance ', pctVarExplained[2]*100))

# plot the loadings of the components PC1 and PC2 
chromvarPCtibble <- tibble(hgnc.symbol = sapply(strsplit(rownames(pcloadings), '_'), function (x) x[[3]]), pc1 = pcloadings[, 1], pc2 = pcloadings[, 2] )
ggplot(chromvarPCtibble) + 
  geom_text(mapping=aes(chromvarPCtibble$pc1, chromvarPCtibble$pc2, label=chromvarPCtibble$hgnc.symbol, size=6))

# calculate motif matching scores for differential peaks
motifPWMsubset <- motifSet[names(motifSet) %in% names(devSubset)]
dp.granges <- granges(fragmentCountsDiffPeaks)
motif_ix <- matchMotifs(motifPWMsubset, 
                        dp.granges, 
                        genome = BSgenome.Hsapiens.UCSC.hg38, 
                        bg = "genome",
                        out = "scores") 

dp.tib <- read_tsv(here('extractedData', 'differentialAtacPeaks.tsv'))
stopifnot(all(end(ranges(dp.granges)) == dp.tib$endLocs))  # make sure diffPeaksTibble is aligned with genomic ranges 

# filter diff peaks to smaller set
dp.tib[["TGFb-high-isDiffPeakAtLeastOneRep"]] & dp.tib[["RA-high-isDiffPeakAtLeastOneRep"]]

diffPeakLogInds <- (dp.tib[["TGFb-high-isDiffPeakAtLeastTwoReps"]] | dp.tib[["TGFb-med-isDiffPeakAtLeastTwoReps"]] | dp.tib[["TGFb-low-isDiffPeakAtLeastTwoReps"]] ) | 
                   (dp.tib[["RA-high-isDiffPeakAtLeastTwoReps"]] | dp.tib[["RA-med-isDiffPeakAtLeastTwoReps"]] | dp.tib[["RA-low-isDiffPeakAtLeastTwoReps"]])

dp.subset.tib <- dp.tib[diffPeakLogInds,] %>% dplyr::select(`RA-med-avgFoldchange`, `TGFb-med-avgFoldchange`, `TGFb-and-RA-med-avgFoldchange`, `EtOH-nlDensity-avgNormFragmentCounts` )
dp.subset.tib <- dp.subset.tib %>%
  mutate(lfcTGFb = log2(`TGFb-med-avgFoldchange`),
         lfcRA = log2(`RA-med-avgFoldchange`),
         lfcBoth = log2(`TGFb-and-RA-med-avgFoldchange`),
         multPred = lfcTGFb + lfcRA)
motif_ix.subset <- motif_ix[diffPeakLogInds,]

for (m in colnames(motif_ix.subset)) {
  scores.temp <- motifScores(motif_ix.subset[, colnames(motif_ix.subset) == m])[,1]
  dp.subset.tib[['scores']] <- scores.temp
  name.short <- strsplit(m, "_")[[1]][3]
  p1 <- ggplot(dp.subset.tib, aes(x = multPred, y = lfcBoth, color = scores)) +
    geom_point(size=I(0.20)) +
    ggtitle(name.short) +
    theme_light()

  ggsave(here("plots", "peakScatterWithMotifScores", sprintf("%s.svg", name.short)), plot = p1, device = "svg")
}


