---
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Applications}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Applications

This vignettes covers some of the various applications of chromVAR's bias corrected deviations and Z-scores for motifs or other annotations. First, we will compute those deviations:

```{r, message = FALSE}
library(chromVAR)
library(motifmatchr)
library(SummarizedExperiment)
library(Matrix)
library(ggplot2)
library(BiocParallel)
library(BSgenome.Hsapiens.UCSC.hg19)
register(SerialParam())
data(example_counts, package = "chromVAR")
set.seed(2017)
example_counts <- addGCBias(example_counts, genome = BSgenome.Hsapiens.UCSC.hg19)
counts_filtered <- filterSamples(example_counts, min_depth = 1500, 
                                 min_in_peaks = 0.15, shiny = FALSE)
counts_filtered <- filterPeaks(counts_filtered)
motifs <- getJasparMotifs()
motif_ix <- matchMotifs(motifs, counts_filtered, genome = BSgenome.Hsapiens.UCSC.hg19)
dev <- computeDeviations(object = counts_filtered, 
                                 annotations = motif_ix)
```

## Variability

The first application is simply to compute the variability of each motif or annotation across the cells or samples of interest.  The function `plotVariability` 

```{r}
variability <- computeVariability(dev)
plotVariability(variability, use_plotly = FALSE)
```

## Clustering

We can also use the bias corrected deviations to cluster the samples. The function `getSampleCorrelation` first removes highly correlated annotations and low variability annotations and then computes the correlation between the cells for the remaining annotations.

```{r, fig.width=5}
sample_cor <- getSampleCorrelation(dev)

library(pheatmap)
pheatmap(as.dist(sample_cor), 
         annotation_row = colData(dev), 
         clustering_distance_rows = as.dist(1-sample_cor), 
         clustering_distance_cols = as.dist(1-sample_cor))
```

## Cell / sample similarity

We can also use tSNE for looking at cell similarity.  The function `deviationsTsne` performs tsne and results a data.frame with the results. If running in an interactive session, shiny can be set to TRUE to load up a shiny gadget for exploring parameters. In general, with 100's of cells a perplexity of around 30-50 might make sense. In this example, we have few cells so we are bounded in how high the perplexity can be set-- we will use 10 here. 

```{r}
tsne_results <- deviationsTsne(dev, threshold = 1.5, perplexity = 10, 
                               shiny = FALSE)
```

To plot the results, `plotDeviationsTsne` can be used.  If running in an interactive session or an interactive Rmarkdown document, shiny can be set to TRUE to generate a shiny widget.  Here we will show static results.

```{r}
tsne_plots <- plotDeviationsTsne(dev, tsne_results, annotation = "TEAD3", 
                                   sample_column = "Cell_Type", shiny = FALSE)
tsne_plots[[1]]
tsne_plots[[2]]
```

## Differential accessibility and variability

The `differentialDeviations` function determines whether there is a significant difference between the bias corrected deviations for a given annotation between different groups. The groups can be specified by giving the column name of a column in the colData of the dev object or a vector of group assignments.


```{r}
diff_acc <- differentialDeviations(dev, "Cell_Type")

head(diff_acc)

```

The `differentialVariability` function determines whether there is a significant difference between the variability of any of the annotations between different groups.

```{r}
diff_var <- differentialVariability(dev, "Cell_Type")

head(diff_var)

```


## Motif / kmer similarity

We can also perform tsne for motif similarity rather than cell similarity, by specifying `what = "annotations"` to the `deviationsTsne` function.  

```{r}
inv_tsne_results <- deviationsTsne(dev, threshold = 1.5, perplexity = 8, 
                                    what = "annotations", shiny = FALSE)

ggplot(inv_tsne_results, aes(x = Dim1, y = Dim2)) + geom_point() + 
  chromVAR_theme()

```

## Kmers and sequence specificity of variation

By using kmers as the annnotation, we can use the kmers to identify the precise nucleotides necessary for chromatin accessibility variability.  The `deviationsCovariability` function returns a normalized covariance between the bias corrected deviations of any pair of annotations. We will use 6mers here in the interest of computational time, but in general 7mers yield higher variability and are better starting points for assembling de novo motifs (see next section).  

```{r}
kmer_ix <- matchKmers(6, counts_filtered, genome = BSgenome.Hsapiens.UCSC.hg19)
kmer_dev <- computeDeviations(counts_filtered, kmer_ix)

kmer_cov <- deviationsCovariability(kmer_dev)

plotKmerMismatch("CATTCC",kmer_cov)
```


## De novo kmer assembly

We can use the `assembleKmers` function to build de novo motifs using the kmer deviation results. The function goes iteratively through the most variable kmers and uses the normalized covariance of the bias corrected deviations for closely related kmers to weight each nucleotide in the motif.  

```{r}
de_novos <- assembleKmers(kmer_dev, progress = FALSE) #no progress bar
de_novos
```

We can use the `pwmDistance` function to see how close our de novo motifs match known motifs. `pwmDistance` returns a list with three matrices- 'dist' has the distance between each pair of motifs, 'strand' the strand of the motif for the match, and 'offset' the offset between the motifs. 

```{r}
dist_to_known <- pwmDistance(de_novos, motifs)

closest_match1 <- which.min(dist_to_known$dist[1,])
dist_to_known$strand[1,closest_match1]

library(ggmotif) # Package on github at AliciaSchep/ggmotif. Can use seqLogo alternatively
library(TFBSTools)

# De novo motif
ggmotif_plot(de_novos[[1]])

# Closest matching known
ggmotif_plot(toPWM(reverseComplement(motifs[[closest_match1]]),type = "prob"))

```

In this case, it looks like the last five base pairs of the de novo motif match well with the first five of the known motif. To get a better sense of whether these might represent the same motif, one can compute the deviations for the de novo motif and see if they are correlated with the known. 

## Synergy between motifs

chromVAR includes a function for computing the "synergy" between pairs of annotations/motifs, where synergy is defined as the excess variability of chromatin accessibility for peaks sharing both motifs compared to a random sub-sample of the same size of peaks with one motif (the one with greater variability). 

## Function for computing synergy

```{r}
getAnnotationSynergy(counts_filtered, motif_ix[,c(83,24,20)])
```

The result is a matrix with Z-scores for the variability synergy of each possible pairing.  Note that this function is pretty slow and should only be performed with a limited selection of motifs and not the full collection!  

## Correlation

An extreme synergy score suggests the possibility of cooperative or competitive binding between the factors that bind the motif. However, if the two factors simply tend to be co-expressed or oppositely expressed, then the synergy score could also be extreme.  It can thus be helpful to look at the correlation between the deviations for the two motifs, in particular the correlation between the deviations for the peak sets that only have one or only the other motif.  There is a function to compute this correlation:

```{r}
getAnnotationCorrelation(counts_filtered, motif_ix[,c(83,24,20)])
```

## Session Info

```{r}
Sys.Date()
```

```{r}
sessionInfo()
```
